Documentation/anti-patterns.md

```diff
@@ -0,0 +1,180 @@
+gRPC:
+
+- Always check for error responses from RPCs and handle them appropriately.
+- Use the `status` field of the error response to determine the type of error that
+  occurred.
+- When retrying failed RPCs, use a backoff strategy to avoid
+  overwhelming the server or exacerbating network issues.
+- Avoid using `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`,
+  as these options can introduce race conditions and result in unreliable and
+  difficult-to-debug code.
+
+### Example: Handling errors from an RPC
+
+The following code snippet demonstrates how to handle errors from an RPC in
+gRPC:
+
+```go 
+ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+defer cancel()
+
+res, err := client.MyRPC(ctx, &MyRequest{})
+if err != nil {
+    // Handle the error appropriately,
+    // log it & return an error to the caller, etc.
+    log.Printf("Error calling MyRPC: %v", err)
+    return nil, err
+}
+
+// Use the response as appropriate 
+log.Printf("MyRPC response: %v", res)
+
+```
+
+To determine the type of error that occurred, you can use the status field of
+the error response:
+
+
+```go 
+resp, err := client.MakeRPC(context.Background(), request) 
+if err != nil {
+  status, ok := status.FromError(err) 
+  if ok {
+    // Handle the error based on its status code 
+    if status.Code() == codes.NotFound {
+      log.Println("Requested resource not found")
+    } else {
+      log.Printf("RPC error: %v", status.Message())
+    }
+  } else {
+    //Handle non-RPC errors 
+    log.Printf("Non-RPC error: %v", err)
+  }
+  return
+}        
+
+// Use the response as needed 
+log.Printf("Response received: %v", resp) 
+```
+
+### Example: Using a backoff strategy
+
+
+When retrying failed RPCs, use a backoff strategy to avoid overwhelming the
+server or exacerbating network issues:
+
+
+```go 
+var res *MyResponse
+var err error
+
+// Retry the RPC call a maximum number of times
+for i := 0; i < maxRetries; i++ {
+    // Create a context with a timeout of one second
+    ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+    defer cancel()
+    
+    // Make the RPC call
+    res, err = client.MyRPC(ctx, &MyRequest{})
+    
+    // Check if the RPC call was successful
+    if err == nil {
+        // The RPC was successful, so break out of the loop
+        break
+    }
+    
+    // The RPC failed, so wait for a backoff period before retrying
+    backoff := time.Duration(i) * time.Second
+    log.Printf("Error calling MyRPC: %v; retrying in %v", err, backoff)
+    time.Sleep(backoff)
+}
+
+// Check if the RPC call was successful after all retries
+if err != nil {
+    // All retries failed, so handle the error appropriately
+    log.Printf("Error calling MyRPC: %v", err)
+    return nil, err
+}
+
+// Use the response as appropriate
+log.Printf("MyRPC response: %v", res)
+
+```
+
+
+## Conclusion
+
+The `FailOnNonTempDialError`, `WithBlock`, and `WithReturnConnectionError`
```

Comment:
Please add a link to this doc from these function's docstrings, as recommended in an earlier comment.